File: app/utils/dawproject_python/point.py
from abc import ABC
from lxml import etree as ET
from classes.doubleAdapter import DoubleAdapter


class Point(ABC):
    def __init__(self, time=None):
        self.time = time

    def to_xml(self):
        point_elem = ET.Element(self.__class__.__name__)
        if self.time is not None:
            point_elem.set("time", DoubleAdapter.to_xml(self.time))
        return point_elem

    @classmethod
    def from_xml(cls, element):
        time = element.get("time")
        time = DoubleAdapter.from_xml(time) if time is not None else None
        return cls(time)

File: app/utils/dawproject/timeline/point.ts
import { DoubleAdapter } from "../doubleAdapter";
import type { IPoint } from "../types";
import { XmlObject } from "../XmlObject";

export abstract class Point extends XmlObject implements IPoint {
  time: number;

  constructor(time?: number) {
    super();
    // Make time optional for deserialization, fromXmlObject will set it
    this.time = time || 0; // Provide a default placeholder
  }

  toXmlObject(): any {
    // Point is abstract, so it doesn't return a root element itself.
    // Subclasses will wrap these attributes in their specific root tag.
    const attributes: any = {};
    if (this.time !== undefined) {
      attributes["@_time"] = DoubleAdapter.toXml(this.time) || "";
    }
    return attributes;
  }

  fromXmlObject(xmlObject: any): this {
    // validate and populate required time attribute
    if (!xmlObject["@_time"]) {
      throw new Error("Required attribute 'time' missing in XML");
    }
    const timeValue = DoubleAdapter.fromXml(xmlObject["@_time"]);
    if (timeValue === undefined) {
      throw new Error("Invalid time value in XML");
    }
    this.time = timeValue;

    return this;
  }
}
